{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "66158277-a812-4f93-8bf6-551d5f49e10d",
   "metadata": {},
   "source": [
    "## Manually get _public_ targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "5dc344c3-b8f5-409b-84ca-411ecade494b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This was copypasted from the site\n",
    "\n",
    "targets = '''NUDT7A\n",
    "Open SGC summary\n",
    "ATAD\n",
    "BRD1A\n",
    "Open SGC summary\n",
    "DCP2B\n",
    "FAM83BA\n",
    "Open SGC summary\n",
    "MURD\n",
    "NUDT4A\n",
    "Open SGC summary\n",
    "OXA10OTA\n",
    "PARP14A\n",
    "Open SGC summary\n",
    "PHIPA\n",
    "PTP1B\n",
    "SMTGR\n",
    "ACVR1A\n",
    "ATAD2A\n",
    "CAMK1DA\n",
    "DCLRE1AA\n",
    "Open SGC summary\n",
    "FALZA\n",
    "Open SGC summary\n",
    "HAO1A\n",
    "Open SGC summary\n",
    "MUREECA\n",
    "NUDT21A\n",
    "NUDT4\n",
    "NUDT5A\n",
    "NUDT7A_CRUDE\n",
    "smTGRNEW\n",
    "STAG1A\n",
    "TBXTA\n",
    "VIM2\n",
    "XX02KALRNA\n",
    "TNCA\n",
    "ALAS2A\n",
    "EPB41L3A\n",
    "mArh\n",
    "INPP5DA\n",
    "nsp13\n",
    "Mac1\n",
    "Mpro\n",
    "NSP15_B\n",
    "MUREECOLI\n",
    "PGN_RS02895PGA\n",
    "CD44MMA\n",
    "Nprot\n",
    "macro-combi\n",
    "NSP16\n",
    "NSP14\n",
    "MID2A\n",
    "SOS3ATA'''.replace('Open SGC summary', '').split()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "980ae614-910f-46a1-ad06-dd10aa706855",
   "metadata": {},
   "source": [
    "## Blast for seq to Uniprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b802d13-197b-41d1-92cd-878297753ba7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note that the first hsp in the alignment.hsps may not be correct\n",
    "# identical sequences may be merged into a single header (pipe seperated)\n",
    "\n",
    "import functools, re, operator, io\n",
    "from typing import Set, List, Dict, Union, Tuple\n",
    "import pandas as pd\n",
    "import requests\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import Record\n",
    "from fragalysis_api import GetTargetsData\n",
    "\n",
    "# get sequence\n",
    "ids: Dict[str, int] = {}\n",
    "seqs: Dict[str, str] = {}\n",
    "for target_name in targets:\n",
    "    ts = GetTargetsData()\n",
    "    ts.set_target_name_url(target_name)\n",
    "    ts.get_target_json()\n",
    "    ids[target_name] = ts.target_json['results'][0]['id']\n",
    "    seqs[target_name] = ts.target_json['results'][0]['sequences'][0]['sequence']\n",
    "    \n",
    "# get uniprot id\n",
    "alignments: Dict[str, Record.Alignment] = {}\n",
    "ranges: Dict[str, Tuple[int, int]] = {}\n",
    "for target_name, seq in seqs.items():\n",
    "    if not seq:\n",
    "        continue\n",
    "    if target_name in alignments:\n",
    "        continue\n",
    "    blast: io.StringIO = NCBIWWW.qblast(program='blastp', \n",
    "                                        database='swissprot', #'refseq_protein', \n",
    "                                        sequence=seq,\n",
    "                                       )\n",
    "\n",
    "    blast_record: Record.Blast = list(NCBIXML.parse(blast))[0]\n",
    "    alignment: Record.Alignment = blast_record.alignments[0]\n",
    "    hsp: Record.HSP = alignment.hsps[0]\n",
    "    alignments[target_name] = blast_record.alignments[0]\n",
    "    ranges[target_name] = (hsp.sbjct_start, hsp.sbjct_end)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dce5c67-fe51-44cc-9ce9-ebfa4be8fb65",
   "metadata": {},
   "source": [
    "## Get Uniprot data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "caf413c0-efc2-4087-89d1-690824c038dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get uniprot data\n",
    "uniprot_data: Dict[str, dict] = {}\n",
    "for target_name, alignment in alignments.items():\n",
    "    uniprot_id: str = re.match(r'.*?\\|([^|]*)\\.\\d+\\|', alignment.title).group(1)\n",
    "    response = requests.get(f'https://rest.uniprot.org/uniprotkb/{uniprot_id}.json')\n",
    "    response.raise_for_status()\n",
    "    uniprot_data[target_name] = response.json()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "348c236a-c3c7-48f1-a5ab-5729ee0f344f",
   "metadata": {},
   "source": [
    "## get count mols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "bd01c41a-32d9-4bd1-b2ee-5480d688d7c5",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "# The fragalysis-api is utterly non-standard and prints stuff\n",
    "\n",
    "from fragalysis_api import GetMoleculesData\n",
    "from IPython.display import display, clear_output\n",
    "\n",
    "mol_data: Dict[str, dict] = {}\n",
    "mol_counts: Dict[str, int] = {}\n",
    "for target_name in ids:\n",
    "    search = GetMoleculesData()\n",
    "    search.set_target_id(target_name)\n",
    "    search.set_molecule_url()\n",
    "    search.set_mol_data()\n",
    "    mol_counts[target_name] = search.get_mol_data['count']\n",
    "    mol_data[target_name] = search.get_mol_data['results']\n",
    "    clear_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38adceb6-04ce-43b4-b422-28713632b803",
   "metadata": {},
   "source": [
    "## combine into table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "83e22cfe-4033-43ac-9103-1770846a36c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_inner = lambda key: {t: d.get(key, None) for t, d in uniprot_data.items()}\n",
    "get_org_inner = lambda key: {t: d['organism'].get(key, None) for t,d in uniprot_data.items()}\n",
    "get_fun_inner = lambda fun: {t: fun(d) for t, d in uniprot_data.items()}\n",
    "\n",
    "def get_recommended(d):\n",
    "    try:\n",
    "        return d['proteinDescription']['recommendedName']['fullName']['value']\n",
    "    except Exception:\n",
    "        return ''\n",
    "    \n",
    "def get_alternativeNames(d):\n",
    "    try:\n",
    "        return tuple([e['fullName']['value'] for e in d['proteinDescription']['alternativeNames']])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_gene(d):\n",
    "    try:\n",
    "        return d['genes'][0]['geneName']['value']\n",
    "    except Exception:\n",
    "        return ''\n",
    "    \n",
    "    \n",
    "def get_alt_genes(d):\n",
    "    try:\n",
    "        return tuple([e['value'] for e in d['genes'][0]['synonyms']])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_ec(d):\n",
    "    try:\n",
    "        return tuple([e['value'] for e in d['proteinDescription']['recommendedName']['ecNumbers']])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_comment_fun(d):\n",
    "    try:\n",
    "        \n",
    "        return tuple([t['value'] for c in d['comments'] if c['commentType'] == 'FUNCTION' for t in c['texts'] ])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_comment_cat(d):\n",
    "    try:\n",
    "        return tuple([c['reaction']['name']  for c in d['comments'] if c['commentType'] == 'CATALYTIC ACTIVITY' ])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_comment_inter(d):\n",
    "    try:\n",
    "        return tuple([t['value'] for c in d['comments'] if c['commentType'] == 'SUBUNIT' for t in c['texts']])\n",
    "    except Exception:\n",
    "        return ()\n",
    "    \n",
    "def get_domains():\n",
    "    descriptions = {}\n",
    "    for t, d in uniprot_data.items():\n",
    "        span = ranges[t]\n",
    "        domains = []\n",
    "        for feat in d['features']:\n",
    "            if feat['type'] != 'Domain':\n",
    "                continue\n",
    "            start = feat['location']['start']['value']\n",
    "            end = feat['location']['end']['value']\n",
    "            if not(ranges[t][0] <= end and ranges[t][1] >= start):\n",
    "                continue\n",
    "            domains.append((feat['description'], end-start))\n",
    "        descriptions[t] = tuple(map(operator.itemgetter(0), sorted(domains, key=lambda d: d[1])))\n",
    "    return descriptions\n",
    "\n",
    "targets = pd.DataFrame(dict(target_id=ids, sequence_length={t: len(s) for t, s in seqs.items()},\n",
    "                            N_hits=mol_counts,\n",
    "                              uniprot_ranges=ranges,\n",
    "                              uniprot_id=get_inner('primaryAccession'),\n",
    "                              #uniprot_acc=get_inner('primaryAccession'),                \n",
    "                              protein_name=get_fun_inner(get_recommended),\n",
    "                              gene_name=get_fun_inner(get_gene),\n",
    "                              organism_scientific_name=get_org_inner('scientificName'),   \n",
    "                              organism_common_name=get_org_inner('commonName'),  \n",
    "                              organism_taxon_id=get_org_inner('taxonId'),\n",
    "                              protein_alt_names=get_fun_inner(get_alternativeNames),\n",
    "                              alt_gene_names=get_fun_inner(get_alt_genes),\n",
    "                              domains=get_domains(),\n",
    "                              catalysis=get_fun_inner(get_comment_cat),\n",
    "                              EC_numbers=get_fun_inner(get_ec),\n",
    "                              function=get_fun_inner(get_comment_fun),\n",
    "                              interaction=get_fun_inner(get_comment_inter),\n",
    "                             ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "98b05d63-312e-4e8f-bd81-efd7002d273b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle, json\n",
    "\n",
    "with open('targets.fasta', 'w') as fh:\n",
    "    for t, s in seqs.items():\n",
    "        fh.write(f'>{t}\\n{s}\\n')\n",
    "        \n",
    "with open('targets_blast.p', 'wb') as fh:\n",
    "    pickle.dump(alignments, fh)\n",
    "    \n",
    "with open('targets_mol.p', 'wb') as fh:\n",
    "    pickle.dump(mol_data, fh)\n",
    "    \n",
    "targets.to_pickle('targets_df.p')\n",
    "targets.to_csv('targets.csv')\n",
    "targets.to_markdown('targets.md')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
